server/datastore/mysql/apple_mdm.go

```diff
@@ -231,3 +235,281 @@ WHERE
+		}
+
+		n, err := res.RowsAffected()
+		if err != nil {
+			return ctxerr.Wrap(ctx, err, "ingest mdm apple hosts from dep sync rows affected")
+		}
+		resCount = n
+
+		// get new host ids
+		args = []interface{}{}
+		parts := []string{}
+		for _, d := range filteredDevices {
+			args = append(args, d.SerialNumber)
+			parts = append(parts, "?")
+		}
+		var hostIDs []uint
+		err = sqlx.SelectContext(ctx, tx, &hostIDs, fmt.Sprintf(`
+			SELECT id FROM hosts WHERE hardware_serial IN(%s)`,
+			strings.Join(parts, ",")),
+			args...)
+		if err != nil {
+			return ctxerr.Wrap(ctx, err, "ingest mdm apple host get host ids")
+		}
+
+		if err := upsertMDMAppleHostDisplayNamesDB(ctx, tx, hostIDs...); err != nil {
+			return ctxerr.Wrap(ctx, err, "ingest mdm apple host upsert related tables")
+		}
+
+		if err := upsertMDMAppleHostLabelMembershipDB(ctx, tx, hostIDs...); err != nil {
+			return ctxerr.Wrap(ctx, err, "ingest mdm apple host upsert related tables")
+		}
+
+		return nil
+	})
+
+	return resCount, err
+}
+
+func upsertMDMAppleHostDisplayNamesDB(ctx context.Context, tx sqlx.ExtContext, hostIDs ...uint) error {
+	args := []interface{}{}
+	parts := []string{}
+	for _, id := range hostIDs {
+		args = append(args, id)
+		parts = append(parts, "(?, '')")
+	}
+
+	_, err := tx.ExecContext(ctx, fmt.Sprintf(`
+			INSERT INTO host_display_names (host_id, display_name) VALUES %s
+			ON DUPLICATE KEY UPDATE display_name = VALUES(display_name)`, strings.Join(parts, ",")),
+		args...)
+	if err != nil {
+		return ctxerr.Wrap(ctx, err, "upsert host display names")
+	}
+
+	return nil
+}
+
+func upsertMDMAppleHostLabelMembershipDB(ctx context.Context, tx sqlx.ExtContext, hostIDs ...uint) error {
+	// Builtin label memberships are usually inserted when the first distributed query results are
+	// received; however, we want to insert pending MDM hosts now because it may still be some time
+	// before osquery is running on these devices. Because these are Apple devices, we're adding
+	// them to the "All Hosts" and "macOS" labels.
+	labelIDs := []uint{}
+	err := sqlx.SelectContext(ctx, tx, &labelIDs, `SELECT id FROM labels WHERE label_type = 1 AND (name = 'All Hosts' OR name = 'macOS')`)
+	switch {
+	case err != nil:
+		return ctxerr.Wrap(ctx, err, "get builtin labels")
+	case len(labelIDs) != 2:
+		// Builtin labels can get deleted so it is important that we check that they still exist
+		// before we continue.
+		return ctxerr.New(ctx, fmt.Sprintf("expected 2 builtin labels but got %d", len(labelIDs)))
+	default:
+		// continue
+	}
+
+	parts := []string{}
+	args := []interface{}{}
+	for _, id := range hostIDs {
+		parts = append(parts, "(?,?),(?,?)")
+		args = append(args, id, labelIDs[0], id, labelIDs[1])
+	}
+	_, err = tx.ExecContext(ctx, fmt.Sprintf(`
+			INSERT INTO label_membership (host_id, label_id) VALUES %s 
+			ON DUPLICATE KEY UPDATE host_id = host_id`, strings.Join(parts, ",")), args...)
```

Comment:
why `DUPLICATE KEY UPDATE host_id = host_id`? I understand doing `ON DUPLICATE KEY UPDATE updated_at = VALUES(updated_at)` like `AsyncBatchInsertLabelMembership` does.

related: should we abstract/reuse `AsyncBatchInsertLabelMembership`? seems like we're doing exactly the same here

Reply:
My goal with `host_id = host_id` was for the duplicate to be noop. I didn't want to use `INSERT IGNORE` since it ignores all errors not just duplicates.

It would be nice to have some agreed upon abstractions for these sorts of operations. Not sure if we should try to tackle that as part of this PR as I think the async functions were built specifically for how we're using redis. I'm definitely in favor of opening an issue to look into it. Or if you think it is an easy change, we could potentially use this PR.